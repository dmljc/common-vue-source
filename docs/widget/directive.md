<!-- ## btnTrack
##### 用于Hubble按钮埋点

<component-wrap>
<el-button v-btnTrack type="primary">按钮埋点</el-button>
</component-wrap>

::: code
```
<el-button v-btnTrack type="primary">按钮埋点</el-button>
```
::: -->

## v-int
##### 仅允许输入数字

<component-wrap>
<div>
    <el-input v-model="input1" placeholder="请输入整数" v-int />
    <p>输入值：{{ input1 }}</p>
</div>
</component-wrap>

::: code
```
<el-input v-model="input1" placeholder="请输入整数" v-int />
```
::: 

## v-float
##### 仅允许输入小数，v-float:num 可指定num位小数

<component-wrap>
<div>
    <el-row :gutter="20">
        <el-col :span="12">
            <el-input v-model="input2" placeholder="请输入数字，最多两位小数" v-float />
            <p>默认最多两位小数，输入值：{{ input2 }}</p>
        </el-col>
        <el-col :span="12">
            <el-input v-model="input3" placeholder="请输入数字，最多四位小数" v-float:4 />
            <p>最多四位小数，输入值：{{ input3 }}</p>
        </el-col>
    </el-row>
</div>
</component-wrap>

::: code
```
<el-input v-model="input2" placeholder="请输入数字，最多两位小数" v-float />

<el-input v-model="input3" placeholder="请输入数字，最多四位小数" v-float:4 />
```
:::

## v-intArr
##### 仅允许输入数字，且可输入多个，逗号、空格分隔，可结合Util.blankToComma使用

<component-wrap>
<div>
    <el-input v-model="input4" placeholder="请输入整数" v-intArr />
    <p>使用 Util.blankToComma：{{ formatInput }}</p>
</div>
</component-wrap>

::: code
```
<el-input v-model="input4" placeholder="请输入整数" v-intArr />
```
::: 

## v-numberChar
##### 仅允许输入数字和字母

<component-wrap>
<div>
    <el-input v-model="input5" placeholder="请输入整数或数字" v-numberChar />
    <p>输入值：{{ input5 }}</p>
</div>
</component-wrap>

::: code
```
<el-input v-model="input5" placeholder="请输入整数或数字" v-numberChar />
```
::: 

<script>
import Util from '@/widget/util';
export default {
    data() {
        return {
            Util,
            input1: '',
            input2: '',
            input3: '',
            input4: '',
            input5: ''
        }
    },
    computed: {
        formatInput() {
            return Util.blankToComma(this.input4);
        }
    },
}
</script>