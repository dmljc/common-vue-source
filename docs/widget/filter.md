## date
##### 用于格式化日期，默认YYYY-MM-DD

<component-wrap>
<div>
    <el-date-picker v-model="now" />
    <p>
        默认：{{ now | date }}
    </p>
</div>
</component-wrap>

::: code
```
<span>{{ now | date }}</span>
```
:::

## datetime
##### 用于格式化日期时间，默认YYYY-MM-DD HH:mm:ss

<component-wrap>
<div>
    <el-date-picker v-model="now" />
    <p>
        默认：{{ now | datetime }}
    </p>
</div>
</component-wrap>

::: code
```
<span>{{ now | datetime }}</span>
```
:::

## number
##### 用于格式化数字格式化，默认保留小数点后2位

<component-wrap>
<div>
    <el-input v-model="input" placeholder="请输入数字" />
    <p>
        默认：{{ input | number }}
    </p>
    <p>
        整数：{{ input | number(0) }}
    </p>
    <p>
        保留四位小数：{{ input | number(4) }}
    </p>
</div>
</component-wrap>

::: code
```
默认：{{ input | number }}
整数：{{ input | number(0) }}
保留四位小数：{{ input | number(4) }}
```
:::






## lsTraySizeStr
##### 用于物流业务中需要将托盘尺寸固定的数据结构在详情页中呈现的场景

<component-wrap>
<span>
    {{traySizeList | lsTraySizeStr}}
</span>
</component-wrap>

:::code
```
<!-- traySizeList: [{length, width, height, count}, ...] -->
<span>{{traySizeList | lsTraySizeStr}}</span>
```
:::

## lsContainerTypeStr
##### 用于物流业务中需要将柜型固定的数据结构在详情页中呈现的场景

<component-wrap>
<span>
    {{containerTypeList | lsContainerTypeStr}}
</span>
</component-wrap>

:::code
```
<!-- containerTypeList: [{value, name}, ...] -->
<span>{{containerTypeList | lsContainerTypeStr}}</span>
```
:::

## lsSelectionNames
##### 用于业务中需要将多选下拉框的id映射为name在详情中展示的场景

<component-wrap>
<span>
    ids: '1,2', names: {{'1,2' | lsSelectionNames(source)}}
    <br />
    idList: [3, 4], names: {{[3, 4] | lsSelectionNames(source)}}
</span>
</component-wrap>

::: code
```
<!-- source: [{id, name}, ...] -->
<span>
    ids: '1,2', names: {{'1,2' | lsSelectionNames(source)}}
    <br />
    idList: [3, 4], names: {{[3, 4] | lsSelectionNames(source)}}
</span>
```
:::

## lsSecondStr
##### 用于业务中需要将秒数呈现为时分秒的场景

<component-wrap>
<span>
    {{3600 | lsSecondStr}}
</span>
</component-wrap>

::: code
```
<span>
    {{3600 | lsSecondStr}}
</span>
```
:::

## lsCascaderNames
##### 用于业务中需要把级联的id映射为name在详情中展示的场景

<component-wrap>
<span>
    idList: [1, 2, 3], names: {{[1, 2, 3] | lsCascaderNames(cascaderSource)}}
</span>
</component-wrap>

::: code
```
<span>
    idList: [1, 2, 3], names: {{[1, 2, 3] | lsCascaderNames(cascaderSource)}}
</span>
```
:::

<script>
export default {
    data() {
        return {
            now: +new Date(),
            input: '',
            traySizeList: [
                {length: 1, width: 2, height: 3, count: 4}, 
                {length: 99, width: 99, height: 99, count: 99}
            ],
            containerTypeList: [
                {value: 1, name: '20GP'},
                {value: 2, name: '40GP'},
                {value: 3, name: '20RF'},
                {value: 4, name: '40RF'},
                {value: 5, name: '40HC'},
                {value: 6, name: '45HC'}
            ],
            source: [
                {id: 1, name: '中国'},
                {id: 2, name: '美国'},
                {id: 3, name: '巴西'},
                {id: 4, name: '英国'}
            ],
            cascaderSource: [
                {   
                    "id": 1,
                    "name":"中国",
                    "children":[
                        {
                            "id": 2,
                            "name":"浙江",
                             "children":[
                                {
                                    "children": null,
                                    "id": 3,
                                    "name": "杭州"
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    }
}
</script>