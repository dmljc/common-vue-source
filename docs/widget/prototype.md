## $audit
##### 用在审核场景直接调用，不需要再写弹窗以及审核逻辑，通过type指定通过或驳回，通过callback执行请求等业务逻辑

<component-wrap>
<el-button id="button-audit" type="primary" @click="$audit({type: 'pass', callback: () => {}})">审核通过</el-button>
<el-button type="primary" @click="$audit({type: 'reject', callback: () => {}})">审核驳回</el-button>
</component-wrap>

::: code
```
<el-button type="primary" @click="$audit({type: 'pass', callback: () => {}})">审核通过</el-button>
<el-button type="primary" @click="$audit({type: 'reject', callback: () => {}})">审核驳回</el-button>
```
:::

<table-wrap-methods :list="auditMethods" />


## $formatTime
##### 参见工具函数formatTime


## $number
##### 参见工具函数number


## $export
##### 用于业务中列表页导出excel的场景，该方法为get请求，后端需要在result.resultUrl中返回excel下载链接；post方法请使用$postExport

<component-wrap>
<el-button type="primary" @click="$export('xxx', {extra: 'data'})">导出</el-button>
<control-button type="primary" @click="$export('xxx', {extra: 'data'}, $event)">导出【防重复点击】</control-button>
</component-wrap>

::: code
```
<el-button type="primary" @click="$export('xxx', {extra: 'data'})">导出</el-button>
<control-button type="primary" @click="$export('xxx', {extra: 'data'}, $event)">导出</control-button>
```
:::

<table-wrap-methods :list="exportMethods" />


## $postExport
##### 同$export，请求为post


## $exportByEmail
##### 用于业务中列表页导出excel且导出过程耗时，需要导出到邮箱的场景，该请求为get请求，同$export,唯一区别在于该请求成功后回调不需要打开excel下载链接，而是直接提示导出邮箱成功。

## $postExportByEmail
##### 同$exportByEmail，请求为post


## $scrollToView
##### 用于业务中需要将指定元素滚动到可是界面的场景；使用该方法需确保滚动父元素，默认为#body-main,如果工程中滚动容器已经为#body-main则可以忽略

<component-wrap>
<el-button type="primary" @click="$scrollToView('button-audit', 200, bodyMain)">滚动到审核按钮</el-button>
</component-wrap>

::: code
```
<el-button type="primary" @click="$scrollToView('button-audit', 200, bodyMain)">滚动到审核按钮</el-button>
<!-- 该工程滚动容器为documentElement，所以需要传递第三个参数 -->
const bodyMain = document.documentElement;
```
:::

<table-wrap-methods :list="scrollMethods" />


## $baiduGcApi
##### 百度地址解析api

<component-wrap>
<div>
    <el-input placeholder="请输入地址" v-model="address" class="f-input-inline-xxl" />
    <el-button type="primary" @click="$baiduGcApi(address, addressCallback)">百度地址解析</el-button>
    <p>{{baiduAddress}}</p>
</div>
</component-wrap>

::: code
```
<div>
    <el-input placeholder="请输入地址" v-model="address" class="f-input-inline-xxl" />
    <el-button type="primary" @click="$baiduGcApi(address, addressCallback)">百度地址解析</el-button>
    <p>{{baiduAddress}}</p>
</div>
function addressCallback(res) {
    this.baiduAddress = JSON.stringify(res);
}
```
:::
<table-wrap-methods :list="baiduMethods" />

<script>
    export default {
        data() {
            return {
                auditMethods: [
                    {name: '$audit', desc: '审核通过或驳回', params: 'options：{type, callback}'}
                ],
                exportMethods: [
                    {name: '$export', desc: '列表页导出excel, 该方法将下载excel到本地', params: 'url: 后端接口地址；params: 请求参数；$event: 使用control-button时需要提供，可用于组件防止按钮重复提交'}
                ],
                scrollMethods: [
                    {name: '$scroll', desc: '滚动到指定元素', params: 'el: 滚动的目标元素，可传递id或dom元素；top: 目标元素距离顶部的距离，一般工程头部占据56px,所以默认值为56px；scrollBody: 滚动容器，工程中一般默认为#body-main,如若不是，需指定'}
                ],
                baiduMethods: [
                    {name: '$baiduGcApi', desc: '百度地址解析api', params: 'address: 地址字符串；callback: 解析后回调'}
                ],
                bodyMain: '',
                address: '浙江省杭州市滨江区',
                baiduAddress: ''
            }
        },
        mounted() {
            this.bodyMain = document.documentElement;
        },
        methods: {
            addressCallback(res) {
                this.baiduAddress = JSON.stringify(res);
            }
        }
    }
</script>

