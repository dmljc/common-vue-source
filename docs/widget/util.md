## renderHeader
##### 用于业务中表格中表头需要鼠标悬浮提示的场景；设置el-table中render-header属性

<component-wrap>
<div>
    <el-table :data="[{name: '张三', sex: '男', age: 18}]">
        <el-table-column label="姓名" prop="name" :render-header="Util.renderHeader('this is tip', 'el-icon-info')" />
        <el-table-column label="性别" prop="sex" :render-header="Util.renderHeader('this is tip', 'el-icon-info', true)" />
        <el-table-column label="年龄" prop="age" :render-header="Util.renderHeader('this is tip')" />
    </el-table>
</div>
</component-wrap>

::: code
```
<el-table :data="[{name: '张三', sex: '男', age: 18}]">
    <el-table-column label="姓名" prop="name" :render-header="Util.renderHeader('this is tip', 'el-icon-info')" />
    <el-table-column label="性别" prop="sex" :render-header="Util.renderHeader('this is tip', 'el-icon-info', true)" />
    <el-table-column label="年龄" prop="age" :render-header="Util.renderHeader('this is tip')" />
</el-table>
```
:::

<table-wrap-attrs title="参数说明" :list="renderHeaderAttrs" />


## prefixInteger
##### 用于公共方法createBelongMonth内部调用，亦可以被外部调用；用于将数字补齐到指定位数，比如用于月份中补齐到两位，1 => 01, 2 => 02, 11 => 11

<component-wrap>
<div>
    <el-input v-model="prefixNumber" class="f-input-inline" />
    <el-button type="primary" @click="fixedNumber = Util.prefixInteger(prefixNumber, 2)">补齐到两位</el-button>
    <span>Result: {{fixedNumber || '-'}}</span>
</div>
</component-wrap>

::: code
```
Util.prefixInteger(prefixNumber, 2);
```
:::

<table-wrap-attrs title="参数说明" :list="prefixAttrs" />


## createBelongMonth
##### 用于业务中查询年月份，该函数自动生成年月份的数据源。

<component-wrap>
<div>
    <el-cascader v-model="belongMonth" :options="belongMonthSource" :props="{value: 'id', label: 'name'}" />
    <span>Selected: {{belongMonth || '-'}}</span>
</div>
</component-wrap>

::: code
```
belongMonthSource = Util.createBelongMonth(2015, true);
```
:::

<table-wrap-attrs title="参数说明" :list="belongMonthAttrs" />


## getNameById
##### 通过id获取下拉数据源中对应的name，提供给过滤器lsSelectionNames内部调用，也可以外部调用，用于详情页中后端未返回文本字段，前端可以通过过滤器将id映射为name

<component-wrap>
<div>
    <el-select v-model="yearValue" class="f-select-inline">
         <el-option
            v-for="item in belongMonthSource"
            :key="item.id"
            :label="item.name"
            :value="item.id" />
    </el-select>
    <span class="f-mr20">Selected Id: {{yearValue || '-'}}</span>
    <span>Selected Name: {{Util.getNameById(yearValue, belongMonthSource) || '-'}}</span>
</div>
</component-wrap>

::: code
```
Util.getNameById(yearValue, belongMonthSource)
```
:::

<table-wrap-attrs title="参数说明" :list="selectionAttrs" />

## download
##### 用于业务中需要下载图片或文件的场景，该函数内部处理了参数拼接以及下载的逐级兼容和退化逻辑，尽可能保证下载的行为；

<component-wrap>
<el-button type="primary" @click="Util.download('https://haitao.nos.netease.com/5b9c52f0-91b2-4205-96ad-7fbb45a01bb0_752_676.png')">下载图片</el-button>
</component-wrap>

::: code
```
<el-button type="primary" @click="Util.download('https://haitao.nos.netease.com/5b9c52f0-91b2-4205-96ad-7fbb45a01bb0_752_676.png')">下载图片</el-button>
```
:::

<table-wrap-attrs title="参数说明" :list="downloadAttrs" />


## datetimeFormat
##### 用于业务中将时间戳格式化为时间格式的场景

<component-wrap>
<span>
    1572867968451 = {{ Util.datetimeFormat(1572867968451) }}
</span>
</component-wrap>

::: code
``` 
<span>
    1572867968451 = {{ Util.datetimeFormat(1572867968451) }}
</span>
```
:::

<table-wrap-attrs title="参数说明" :list="dateTimeAttrs" />

## numberFormat
##### 用于格式化数字，小数点后默认保留 2 位小数

<component-wrap>
<span>
    Util.numberFormat(0.1) = {{ Util.numberFormat(0.1) }}
</span>
</component-wrap>

::: code
```
<span>
    Util.numberFormat(0.1) = {{ Util.numberFormat(0.1) }}
</span>
```
:::

<table-wrap-attrs title="参数说明" :list="numberFormatAttrs" />

## str2arr
##### 用于将类数组形式的字符串分割为数组

<component-wrap>
<span>
    Util.str2arr('1,2,3') = {{ Util.str2arr('1,2,3')}}<br />
    Util.str2arr('1-2-3', '-') = {{ Util.str2arr('1-2-3', '-')}}<br />
    Util.str2arr('1-2-3', '-', true) = {{ Util.str2arr('1-2-3', '-', true)}}
</span>
</component-wrap>

::: code
``` html
<span>
    Util.str2arr('1,2,3') = {{ Util.str2arr('1,2,3')}}<br />
    Util.str2arr('1-2-3', '-') = {{ Util.str2arr('1-2-3', '-')}}<br />
    Util.str2arr('1-2-3', '-', true) = {{ Util.str2arr('1-2-3', '-', true)}}
</span>
```
:::

<table-wrap-attrs title="参数说明" :list="str2arrAttrs" />

## filterEmpty
##### 用于过滤对象中 value 为 undefined, '', null 等无效值；如果想要保留对象中的空字符，可以在对象上增加一个属性_allowSpace为true。

<component-wrap>
<span>
    {{ Util.filterEmpty(filterObj)}}
</span>
</component-wrap>

::: code
``` html
<!-- filterObj: {
    a: 1,
    b: '',
    c: null,
    d: undefined,
    e: 0,
    f: false,
    g: true,
    h: NaN
} -->
{{ Util.filterEmpty(filterObj)}}
```
:::

<table-wrap-attrs title="参数说明" :list="filterEmptyAttrs" />

<script>
    import Util from '@/widget/util';
    export default {
        data() {
            const belongMonthSource = Util.createBelongMonth(2015, true);
            return {
                Util,
                renderHeaderAttrs: [
                    {name: 'content', desc: '鼠标悬浮提示的文字', type: 'String', optional: 'Any String', default: ''},
                    {name: 'icon', desc: '图标，不指定则没有图标', type: 'String', optional: 'element icon图标或项目iconfont', default: ''},
                    {name: 'required', desc: '是否带必填*标志', type: 'Boolean', optional: 'Boolean', default: 'false'}
                ],
                prefixNumber: 1,
                fixedNumber: '',
                prefixAttrs: [
                    {name: 'num', desc: '被格式化的数字', type: 'Number', optional: 'Number', default: ''},
                    {name: 'length', desc: '格式化后的长度', type: 'Number', optional: 'Any Number', default: '2'}
                ],
                belongMonth: [],
                belongMonthSource,
                belongMonthAttrs: [
                    {name: 'start', desc: '数据源的开始年份', type: 'Number', optional: 'year Number', default: '2014'},
                    {name: 'prefixMonth', desc: '是够对选中的值进行月份补全,取决于后端期望的数据格式', type: 'Boolean', optional: 'Boolean', default: 'false'}
                ],
                yearValue: '',
                selectionAttrs: [
                    {name: 'id', desc: '数据id', type: 'Number | String', optional: 'Number | String', default: ''},
                    {name: 'source', desc: '数据源', type: 'Array', optional: 'Array', default: '[]'},
                ],
                downloadAttrs: [
                    {name: 'url', desc: '下载链接', type: 'String', optional: 'url String', default: ''},
                    {name: 'params', desc: '下载额外的参数', type: 'Object', optional: 'Object', default: ''},
                    {name: 'isImage', desc: '是否图片，该字段会影响下载逻辑；如果非图片，下载行为会直接通过window.open进行下载；', type: 'Boolean', optional: 'Boolean', default: 'true'}
                ],
                dateTimeAttrs: [
                    {name: 'timestamp', desc: '被格式化的时间戳', type: 'Number', optional: 'Number', default: ''},
                    {name: 'format', desc: '格式化后的时间格式', type: 'String', optional: 'String', default: 'YYYY-MM-DD HH:mm:ss'}
                ],
                numberFormatAttrs: [
                    {name: 'value', desc: '被格式化的数字', type: 'Number', optional: 'Number', default: ''},
                    {name: 'keep', desc: '保留的小数点位数', type: 'Number', optional: 'Number', default: '2'}
                ],
                str2arrAttrs: [
                    {name: 'str', desc: '被转化的字符串', type: 'String', optional: 'String', default: ''},
                    {name: 'split', desc: '字符串连接符', type: 'String', optional: 'String', default: ','},
                    {name: 'parse', desc: '分割为数组后是否对值进行parseInt转换为Number', type: 'Boolean', optional: 'Boolean', default: 'false'}
                ],
                filterEmptyAttrs: [
                    {name: 'filterObj', desc: '被转化的对象', type: 'Object', optional: 'Object', default: ''}
                ],
                filterObj: {
                    a: 1,
                    b: '',
                    c: null,
                    d: undefined,
                    e: 0,
                    f: false,
                    g: true,
                    h: NaN
                }
            }
        }
    }
</script>