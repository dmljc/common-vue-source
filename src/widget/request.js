import axios from 'axios';
import qs from 'qs';
import _ from '@/widget/util';
import { Message } from 'element-ui';

const JSONAPI = axios.create({
    timeout: 60000,
    headers: {
        'X-Requested-With': 'XMLHttpRequest',
        'Content-Type': 'application/json;charset=utf-8'
    },
    transformRequest: [function(data) {
        return JSON.stringify(_.filterEmpty(data));
    }],
    paramsSerializer(params) {
        return qs.stringify(_.filterEmpty(params));
    }
});

function __responseSuccessInterceptor(response) {
    const data = response.data;
    if (data && data.code >= 200 && data.code < 400) {
        data.message && Message.success(data.message);
        return Promise.resolve(data);
    }

    data.message && Message.error(data.message);
    return Promise.reject(data);
}

function __responseErrorInterceptor(error) {
    if (error.response) {
        Message.error('请求失败');
    }
    return Promise.reject(error);
}

JSONAPI.interceptors.response.use(__responseSuccessInterceptor, __responseErrorInterceptor);

export {
    JSONAPI
};
