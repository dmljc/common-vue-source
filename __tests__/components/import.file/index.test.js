/**
 * @jest-environment jsdom
 */
import { createLocalVue, mount } from '@vue/test-utils';
import Compose from '@/main';
import ImportFile from '@/components/import.file';
import { JSONAPI } from '@/widget/request';

jest.mock('@/widget/request', () => {
    const resolveResult = jest.fn();
    resolveResult.mockReturnValueOnce({
        code: 200,
        data: {
            data: 'url1'
        }
    }).mockReturnValueOnce({
        code: 200,
        data: {
            resultUrl: 'url2'
        }
    }).mockReturnValueOnce({
        code: 200,
        data: {
            url: 'url3'
        }
    }).mockReturnValueOnce({
        code: 200,
        result: {
            data: 'url4'
        }
    }).mockReturnValueOnce({
        code: 200,
        result: {
            resultUrl: 'url5'
        }
    }).mockReturnValueOnce({
        code: 200,
        result: {
            url: 'url6'
        }
    }).mockReturnValueOnce({
        code: 200,
        result: {}
    }).mockReturnValueOnce({
        code: 200
    });

    const fn = () => new Promise((resolve) => {
        resolve(resolveResult());
    });
    return {
        __esModule: true,
        JSONAPI: {
            get: jest.fn(fn),
            post: jest.fn(fn)
        }
    };
});

let customMount;
beforeAll(() => {
    const localVue = createLocalVue();
    localVue.use(Compose);
    customMount = options => mount(ImportFile, Object.assign({ localVue }, options));
});

describe('import-file default props & data', () => {
    it('import-file default props & data', (done) => {
        const wrapper = customMount();
        window.setTimeout(() => {
            expect(wrapper.props('extraData')).toEqual({});
            expect(wrapper.props('onLoadInterceptor')({ code: 200 })).toEqual({ code: 200 });
            expect(wrapper.props('downloadAsync')).toBe(false);
            expect(wrapper.props('onSuccessCb')).toBe(null);
            expect(wrapper.vm.status).toBe('init');
            expect(wrapper.vm.uploadingPercent).toBe(0);
            expect(wrapper.vm.uploadingInterval).toBe('');
            expect(wrapper.vm.resultMessage).toBe('');
            expect(wrapper.vm.resultUrl).toBe('');
            expect(wrapper.vm.files).toEqual([]);
            expect(wrapper.element).toMatchSnapshot();
            done();
        }, 10);
    });
});

describe('import-file customs props', () => {
    it('import-file has correct custom props behavior', (done) => {
        const wrapper = customMount({
            propsData: {
                title: '报表导入',
                accept: '.xlsx,.xls,.img',
                extraData: { flag: 1 },
                templateUrl: '/test/templateUrl',
                importUrl: '/test/importUrl',
                visible: true,
                onLoadInterceptor: () => ({ code: 200 }),
                downloadAsync: true
            }
        });
        window.setTimeout(() => {
            expect(wrapper.element).toMatchSnapshot();
            done();
        }, 10);
    });
});

describe('import-file methods', () => {
    it('method[downloadTemplate] has correct behavior -[downloadAsync]', (done) => {
        const wrapper = customMount({
            propsData: {
                templateUrl: '/test/templateUrl',
                visible: true,
                downloadAsync: true
            }
        });
        window.open = jest.fn();
        window.setTimeout(async () => {
            // 测试按钮下载后端返回的六种结果分支
            const downloadBtn = wrapper.find('.import-init__download');
            await downloadBtn.trigger('click');
            expect(JSONAPI.get).toBeCalledWith('/test/templateUrl');
            expect(window.open).toBeCalledWith('url1', '_self');
            await downloadBtn.trigger('click');
            expect(window.open).toBeCalledWith('url2', '_self');
            await downloadBtn.trigger('click');
            expect(window.open).toBeCalledWith('url3', '_self');
            await downloadBtn.trigger('click');
            expect(window.open).toBeCalledWith('url4', '_self');
            await downloadBtn.trigger('click');
            expect(window.open).toBeCalledWith('url5', '_self');
            await downloadBtn.trigger('click');
            expect(window.open).toBeCalledWith('url6', '_self');
            window.open.mockClear();
            await downloadBtn.trigger('click');
            expect(window.open.mock.calls.length).toBe(0);
            await downloadBtn.trigger('click');
            expect(window.open.mock.calls.length).toBe(0);
            done();
        }, 10);
    });

    it('method[downloadTemplate] has correct behavior -[downloadSync]', (done) => {
        const wrapper = customMount({
            propsData: {
                templateUrl: '/test/templateUrl',
                visible: true,
                downloadAsync: false
            }
        });
        window.open = jest.fn();
        window.setTimeout(async () => {
            const downloadBtn = wrapper.find('.import-init__download');
            await downloadBtn.trigger('click');
            expect(window.open).toBeCalledWith('/test/templateUrl', '_self');
            done();
        }, 10);
    });

    it('method[createFormData] has correct behavior', (done) => {
        const wrapper = customMount({
            propsData: {
                extraData: {a: 'x'}
            }
        });
        window.setTimeout(() => {
            wrapper.vm.files = ['file1', 'file2'];
            const data = wrapper.vm.createFormData();
            expect(data.getAll('file')[0]).toBe('file1');
            expect(data.getAll('file')[1]).toBe('file2');
            expect(data.get('a')).toBe('x');
            done();
        }, 10);
    });
});