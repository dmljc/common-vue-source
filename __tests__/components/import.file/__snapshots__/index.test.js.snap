// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import-file customs props import-file has correct custom props behavior 1`] = `
<div
  class="el-dialog__wrapper import-file-block"
  name="dialog-fade"
  style="z-index: 2001;"
>
  <div
    class="el-dialog"
    style="width: 550px; margin-top: 15vh;"
  >
    <div
      class="el-dialog__header"
    >
      <span
        class="el-dialog__title"
      >
        报表导入
      </span>
      <button
        aria-label="Close"
        class="el-dialog__headerbtn"
        type="button"
      >
        <i
          class="el-dialog__close el-icon el-icon-close"
        />
      </button>
    </div>
    <div
      class="el-dialog__body"
    >
      <div>
        <div
          class="import-init"
        >
          <span>
            选择要导入的文件：
          </span>
           
          <button
            class="el-button el-button--primary el-button--small"
            type="button"
          >
            <!---->
            <!---->
            <span>
              选择文件
            </span>
          </button>
           
          <a
            class="import-init__download"
            href="javascript:;"
            target="_blank"
          >
            模板文件下载
          </a>
            
          <input
            accept=".xlsx,.xls,.img"
            name="file"
            style="display: none;"
            type="file"
          />
        </div>
         
        <div
          class="import-tip"
        >
          <h5>
            注意事项：
          </h5>
           
          <p
            class="import-tip__item"
          >
            1. 请按照导出文件的格式进行导入，系统数据以本次导入的为准，历史数据的会被覆盖掉；
          </p>
           
          <p
            class="import-tip__item"
          >
            2. 单次导入的数据最多5000行，如果数据量大，请分批导入；
          </p>
           
          <p
            class="import-tip__item"
          >
            3. 文件大小不能超过2M。
          </p>
        </div>
      </div>
       
      <!---->
       
      <!---->
       
      <!---->
       
    </div>
    <div
      class="el-dialog__footer"
    >
      <div>
        <button
          class="el-button el-button--default"
          type="button"
        >
          <!---->
          <!---->
          <span>
            取消
          </span>
        </button>
         
        <button
          class="el-button el-button--primary is-disabled"
          disabled="disabled"
          type="button"
        >
          <!---->
          <!---->
          <span>
            开始导入
          </span>
        </button>
         
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`import-file default props & data import-file default props & data 1`] = `
<div
  class="el-dialog__wrapper import-file-block"
  name="dialog-fade"
  style="display: none;"
>
  <div
    class="el-dialog"
    style="width: 550px; margin-top: 15vh;"
  >
    <div
      class="el-dialog__header"
    >
      <span
        class="el-dialog__title"
      >
        导入
      </span>
      <button
        aria-label="Close"
        class="el-dialog__headerbtn"
        type="button"
      >
        <i
          class="el-dialog__close el-icon el-icon-close"
        />
      </button>
    </div>
    <!---->
    <div
      class="el-dialog__footer"
    >
      <div>
        <button
          class="el-button el-button--default"
          type="button"
        >
          <!---->
          <!---->
          <span>
            取消
          </span>
        </button>
         
        <button
          class="el-button el-button--primary is-disabled"
          disabled="disabled"
          type="button"
        >
          <!---->
          <!---->
          <span>
            开始导入
          </span>
        </button>
         
        <!---->
      </div>
    </div>
  </div>
</div>
`;
